### PR TITLE
build-system: improve terser/esbuild integration

### PR DESCRIPTION
**Pull request by samouri**
**summary**
The first in a series of small cleanups to our build helpers. The improvements are extractions from a much larger https://github.com/ampproject/amphtml/pull/35201

This PR
- Updates `esbuild` to write only to memory instead of `fs`
- Removes esbuild minification as it doesn't add much in conjunction with terser. It also makes enabling private prop mangling impossible (which we aren't doing yet....but stay tuned).
- Terser updates
  - Strip comments
  - Operate in module mode depending on `argv.esm`
  - Operate in-mem instead of reading/writing to fs.
  - Use builtin sourcemap conversion instead of deferring to `remapping`
  - Adds a preamble of `;`, like we do with closure.

Next up improvements:
- Combining `compileWithEsbuild` and `compileUnminifiedJs`
- `finishBundle` should also operate in-mem...this may require more refactoring than its worth though

### CODE REVIEW

File: build-system/tasks/helpers.js [hunk 1]
```diff
@@ -602,39 +615,28 @@ async function compileJsWithEsbuild(srcDir, srcFilename, destDir, options) {
 /**
  * Minify the code with Terser. Only used by the ESBuild.
  *
- * @param {string} destDir
- * @param {string} destFilename
- * @param {?Object} options
- * @return {!Promise}
+ * @param {string} code
+ * @param {string} map
+ * @return {!Promise<{code: string, map: *, error?: Error}>}
```
Comment (samouri) [inline build-system/tasks/helpers.js:620]: self-nit: `error?` should be removed

File: build-system/tasks/helpers.js [hunk 2]
```diff
@@ -529,29 +529,42 @@ async function compileJsWithEsbuild(srcDir, srcFilename, destDir, options) {
...
+    if (options.minify) {
+      ({code, map} = await minify(code, map));
+    }
+
+    await Promise.all([
+      fs.outputFile(destFile, code),
+      fs.outputFile(`${destFile}.map`, map),
+    ]);
+
+    // TODO: finishBundle should operate in-mem instead of reading/writing from FS.
```
Comment (rsimha) [inline build-system/tasks/helpers.js:566]: Agree with the sentiment behind this statement. In practice, the in-memory stuff we used to do with `gulp` streams once upon a time was way slower than the current direct-to-file-system operations. Going back to in-memory (without `gulp`) will be worth it if we can achieve yet another order of magnitude speed up.

File: build-system/tasks/helpers.js [hunk 3]
```diff
@@ -602,39 +615,28 @@ async function compileJsWithEsbuild(srcDir, srcFilename, destDir, options) {
...
-  }
-
-  const filename = path.join(destDir, destFilename);
+async function minify(code, map) {
   const terserOptions = {
-    mangle: true,
-    compress: true,
+    mangle: {},
+    compress: {
+      passes: 3,
```
Comment (rsimha) [inline build-system/tasks/helpers.js:626]: Maybe add a comment / link to justify the choice of 3 passes?

File: build-system/tasks/helpers.js [hunk 4]
```diff
@@ -602,39 +615,28 @@ async function compileJsWithEsbuild(srcDir, srcFilename, destDir, options) {
...
-  const remapped = remapping(
-    [minified.map, fs.readFileSync(`${filename}.map`, 'utf8')],
-    () => null,
-    !argv.full_sourcemaps
-  );
-  fs.writeFileSync(filename, minified.code);
-  fs.writeFileSync(`${filename}.map`, remapped.toString());
+
+  const minified = await terser.minify(code, terserOptions);
+  return {code: minified.code ?? '', map: minified.map};
```
Comment (rsimha) [inline build-system/tasks/helpers.js:639]: I assume the nullish coalescing is to satisfy type checking? Could `minified.code` ever realistically be nullish, and if so, are we swallowing an error by defaulting to `''`?

**Review by rsimha — APPROVED** [review]
LGTM with some non-blocking questions / comments. Thank you!

File: build-system/tasks/helpers.js (hunk 3 above)
Comment (samouri) [reply build-system/tasks/helpers.js:626]: Sure. Although the answer isn't too satisfying. It was simply experimentation. Minor improvements until `3`, none after that

File: build-system/tasks/helpers.js (hunk 2 above)
Comment (samouri) [reply build-system/tasks/helpers.js:566]: Thank you for the context! I know we'll need to leave some high level funcs that operate on FS around for test infra, but I can't imagine in-mem being slower for this flow

File: build-system/tasks/helpers.js (hunk 4 above)
Comment (samouri) [reply build-system/tasks/helpers.js:639]: Exactly right. I don't think it could be nullish. If you prefer, I could check for nullish higher up and throw in that case? I'm not sure what to do when the typing is bad...maybe cast?

Comment (rsimha) [reply build-system/tasks/helpers.js:639]: Hmm. I only asked the question to make sure we didn't miss an easy trick. I don't actually have any ideas to make this better than it is. My guess is that any meaningful error during `terser.minify` will be thrown, and we'll never actually silently return an empty string. And even if we do, the very next step in our build pipeline will likely error out with a different meaningful error.

TL;DR: Okay to leave the code as is, given that this discussion now exists for posterity 😃
